cdap-api/src/main/java/co/cask/cdap/api/workflow/WorkflowSpecification.java

```diff
@@ -32,21 +34,28 @@
   private final String className;
   private final String name;
   private final String description;
-  private final Map<String, WorkflowActionSpecification> customActionMap;
-  private final List<ScheduleProgramInfo> actions;
   private final Map<String, String> properties;
 
+  private List<WorkflowNode> nodes = Lists.newArrayList();
+  private Map<String, WorkflowForkSpecification> forks = Maps.newHashMap();
+  private Map<String, ScheduleProgramInfo> actions = Maps.newHashMap();
+  private Map<String, WorkflowActionSpecification> customActionMap = Maps.newHashMap();
+
   public WorkflowSpecification(String className, String name, String description,
-                                      Map<String, String> properties,
-                                      List<ScheduleProgramInfo> actions,
+                                      Map<String, String> properties, List<WorkflowNode> nodes,
+                                      Map<String, WorkflowForkSpecification> forks,
+                                      Map<String, ScheduleProgramInfo> actions,
                                       Map<String, WorkflowActionSpecification> customActionMap) {
     this.className = className;
     this.name = name;
     this.description = description;
     this.properties = properties == null ? Collections.<String, String>emptyMap() :
                                            Collections.unmodifiableMap(new HashMap<String, String>(properties));
-    this.actions = Collections.unmodifiableList(new ArrayList<ScheduleProgramInfo>(actions));
```

Comment:
@chtyim @anew I have restructured the specifications such that - 
1. `Workflow` contains list of `WorkflowNode`
2. `WorkflowNode` is an abstract class which contains `nodeId` and `nodeType`. 
3. `WorkflowNode` is extended by `WorkflowActionNode` which represents the executable node in the `Workflow` and contains `ScheduleProgramInfo`
4. `WorkflowNode` is also extended by `WorkflowForkNode` which represents fork in the `Workflow` and contain list of `WorkflowForkBranch`. Each `WorkflowForkBranch` simply contains the list of `WorkflowNode`.
5. Custom serialization codecs are added for the `WorkflowNode` and `WorkflowForkBranch` to help serializing and deserializing the `WorkflowNode` based on its type.